addons/membership/models/invoice.py

```diff
@@ -0,0 +1,87 @@
+# -*- coding: utf-8 -*-
+from datetime import date
+from openerp import api, models, fields, _
+import openerp.addons.decimal_precision as dp
+from openerp.exceptions import ValidationError
+
+
+class Invoice(models.Model):
+    _inherit = 'account.invoice'
+   
+    @api.multi
+    def action_cancel(self):
+        '''Create a 'date_cancel' on the membership_line object'''
+        MembershipLine = self.env['membership.membership_line']
+        today_date = date.today()
+        for invoice in self:
+            mlines = MembershipLine.search([('account_invoice_line_id', 'in',
+                                            [l.id for l in invoice.invoice_line_ids])])
```

Comment:
instead of doing every time search for membershipline, you can directly do like this way: 

invoice_line_ids = self.mapped('invoice_line_ids').ids
MembershipLine.search([('account_invoice_line_id', 'in', invoice_line_ids)]). no need of for loop.